### DISCUSSION TITLE
B200mini USRP radio from Ettus Research works on WSL2

### DISCUSSION THREAD

**Discussion by bkosanovic** [dorssel/usbipd-win#250] [General]
I was able to get the B200mini USRP radio from Ettus Research to work on WSL2 using usbipd-win v2.0.0.

Device ID = 2500:0021

For this type of devices to work one has to install a few more things on Windows and WSL2.

Windows 11 Pro:

1. WinUSB UHD driver for ettus radios: https://files.ettus.com/binaries/misc/
2. UHD drivers for Windows (I wanted this specific version): https://files.ettus.com/binaries/uhd/uhd_003.015.000.000-release/Windows-10-x64/
3. libusb-1.0 (I took it from here): https://sourceforge.net/projects/libusb/files/libusb-1.0/

Upon getting libusb-1.0, you need to manually copy libusb-1.0.dll to the UHD in C:\Program Files\UHD\bin (or wherever the uhd.dll was placed).

Once you did all of the above, you may start power shell (admin mode) and run uhd_find_devices (make sure you check the .\UHD\bin folder above is in your path)

WSL2 (ubuntu2004):

1. Compile/build/install UHD Driver for ubuntu2004 by following instructions from here: https://kb.ettus.com/Building_and_Installing_the_USRP_Open-Source_Toolchain_(UHD_and_GNU_Radio)_on_Linux
2. Make sure you pick up the same driver version tag as you did for the Windows side (in my case v3.15.0.0).
3. You do not need to install GNU Radio if you are only interested in the UHD driver

Once you have done the above, you can get Windows to recognize your bus series USRP radio (in my case B200mini) and to use it within the WSL2.

To do this you have to make sure you do the following:

1. On windows before you try to use it in WSL2 and before you attach the device to WSL make sure you run uhd_find_devices from the power shell (admin mode). That will load firmware to the radio and get it to be recognized as USB3.0 device in Windows. (by default it is USB2.0)
2. Now, you can attach the device to your running WSL2 instance. Otherwise, the device would appear as USB2 and will disconnect when you run UHD drivers in WSL2 because it would change the BUS ID when switching from USB2 to USB3.

Everything else remains the same.

Enjoy using USRP B series radios in WSL2. Thank you USBIPD-WIN developers and Microsoft for making this possible!

Regards,
Bogdan

**Comment by Dafeigy** [comment]
Hi, I followed your steps and try to connect B210 to WSL2 and run openairinterface. My UHD version is 4.6.0.0., and B210 can be both found in WIN/WSL by using command `uhd_find_devices`. I check the connection between USRP and WINDOWS, and it shows it is USB3.0 device; However, when I use `lsusb` in WSL2 and find these:

```bash
cybersh1t@DESKTOP-8HS1RCL:~/openairinterface$ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 005: ID 2500:0020 Ettus Research LLC USRP B210
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

B210 is still under Bus 001 which is a USB2.0 connection. Any ideas to switch it to 3.0? Since I really need 3.0 speed to obtain better performance. Thanks ^_^

**Reply by raivisr** [reply]
use `uhd_usrp_probe` command, this one loads bitstream/firmware into the device
